pkg/ui/server.go

```diff
@@ -0,0 +1,204 @@
+
+// FormDetails represents all the input and content used for the token retrievlal form.
+type FormDetails struct {
+	WindowName string
+	Origin     string
+	PageTitle  string
+	Content    Content
+	Category   string
+	Reason     string
+	TTL        string
+	Errors     map[string]string
+}
+
+// SuccessDetails represents the data used for the success page and the postMessage response to the client.
+type SuccessDetails struct {
+	Token        string
+	TargetOrigin string
+	WindowName   string
+}
+
+var (
+	categories []string
+	ttls       []string
+)
+
+// RunServer initializes a server on port 9091 and registers a handler for the /popup route.
+func RunServer(ctx context.Context) {
+	categories = []string{"explanation", "breakglass"}
+	ttls = []string{"15", "30", "60", "120", "240"}
+
+	router := http.NewServeMux()
+	server := &http.Server{
+		Addr:         ":9091",
+		Handler:      router,
+		ReadTimeout:  5 * time.Second,
+		WriteTimeout: 1 * time.Second,
+		IdleTimeout:  15 * time.Second,
+	}
+
+	fs := http.FileServer(http.Dir("./assets/static"))
+	router.Handle("/assets/static/", http.StripPrefix("/assets/static/", fs))
+	router.HandleFunc("/popup", popup)
+	log.Fatal(server.ListenAndServe())
+}
+
+func popup(w http.ResponseWriter, r *http.Request) {
+	details := FormDetails{
+		WindowName: r.FormValue("windowname"),
+		Origin:     r.FormValue("origin"),
+		PageTitle:  "JVS - Justification Request System",
+		Category:   r.FormValue("category"),
+		Reason:     r.FormValue("reason"),
+		TTL:        r.FormValue("ttl"),
+		Content: Content{
+			UserLabel:     "User",
+			CategoryLabel: "Category",
+			ReasonLabel:   "Reason",
+			TTLLabel:      "TTL",
+			Categories: []Pair{
+				{
+					Key:  "explanation",
+					Text: "Explanation",
+				},
+				{
+					Key:  "breakglass",
+					Text: "Breakglass",
+				},
+			},
+			TTLs: []Pair{
+				{
+					Key:  "15",
+					Text: "15m",
+				},
+				{
+					Key:  "30",
+					Text: "30m",
+				},
+				{
+					Key:  "60",
+					Text: "1h",
+				},
+				{
+					Key:  "120",
+					Text: "2h",
+				},
+				{
+					Key:  "240",
+					Text: "4h",
+				},
+			},
+		},
+	}
+
+	// Initial page load, just render the page
+	if r.Method == "GET" {
+		// set some defaults
+		details.Category = categories[0]
+		details.TTL = ttls[0]
+		render(w, "./assets/templates/index.html.tmpl", details)
+		return
+	}
+
+	// Form submission
+	if r.Method == "POST" {
+		// 1. Validate input
+		if !details.Validate() {
+			render(w, "./assets/templates/index.html.tmpl", details)
+			return
+		}
+
+		// 2. [TODO] Request a token
+		token := "token_from_server"
+
+		// 3. Redirect to a confirmation page with context, ultimately needed to postMessage back to the client
+		successContext := SuccessDetails{
+			Token:        token,
+			TargetOrigin: details.Origin,
+			WindowName:   details.WindowName,
+		}
+
+		render(w, "./assets/templates/success.html.tmpl", successContext)
+	}
+}
+
+func (formDetails *FormDetails) Validate() bool {
```

Comment:
Stylistically, it would be better to pass FormDetails into a private validate function than to have a method on the details itself.

Reply:
Ack